### PR TITLE
Add initial implementation of REST helpers

### LINKED ISSUE #384: Add helpers for form rendering

### PR DESCRIPTION
**Pull request by jerben**
Fix #384 #447 #383

### CODE REVIEW

**Comment by jerben** [comment]
This should be the bigger part of 0.5.0, tested with `sihl-demo`. The next step is to generate the default view using only the schema :)

File: sihl/src/sihl.mli [hunk 1]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+        -> ('a, 'b, t) Conformist.t
+        -> Rock.Request.t
+        -> Rock.Response.t Lwt.t
+
+      (** [delete name request] handles the deletion of a resource instance and
+          returns the deletion result as a response.
+
+          [name] is the name of the resource in plural, for example [orders] or
+          [users]. *)
+      val delete' : string -> Rock.Request.t -> Rock.Response.t Lwt.t
```
Comment (aronerben) [inline sihl/src/sihl.mli:358]: Can we find a name that does not need a tick?

File: sihl/src/web_rest.ml [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+;;
+
+let routers_of_actions
+    (type a)
+    name
+    schema
+    (module Controller : CONTROLLER with type t = a)
+    (actions : action list)
+  =
+  let rec create_routers (actions : action list) (routers : Web.router list) =
```
Comment (aronerben) [inline sihl/src/web_rest.ml:314]: Can this not be replaced with a `List.map (...) actions`?

**Review by aronerben — CHANGES_REQUESTED** [review]
Please check, clean implementation though

File: sihl/src/index.mld [hunk 3]
```diff
@@ -505,6 +505,128 @@ Logs.info (fun m -> m "This allows the program user to understand what is going
...
+
+With {!Sihl.Web.Rest} you can make apps and services accessible through the web. A resource [pizzas] can have following HTTP routes. Each route is associated to an action.
+
+{v
+GET       /pizzas          `Index  display a list of all pizzas
+GET       /pizzas/new      `New    return an HTML form for creating a new pizza
+POST      /pizzas          `Create create a new pizza
+GET       /pizzas/:id      `Show   display a specific pizza
+GET       /pizzas/:id/edit `Edit   return an HTML form for editing a pizza
+PATCH/PUT /pizzas/:id      `Update a specific pizza
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/index.mld:522]: Fixed.

File: sihl/src/index.mld [hunk 4]
```diff
@@ -505,6 +505,128 @@ Logs.info (fun m -> m "This allows the program user to understand what is going
...
+DELETE    /pizzas/:id      `Delete a specific pizza
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/index.mld:523]: Fixed.

File: sihl/src/index.mld [hunk 5]
```diff
@@ -505,6 +505,128 @@ Logs.info (fun m -> m "This allows the program user to understand what is going
...
+{3 Schema}
+
+A schema connects the static world of OCaml types and the dynamic world of forms and urlencoded values. Sihl uses {{:https://github.com/oxidizing/conformist} Conformist} to decode and validate data.
+
+Attach your own validators to schema fields to validate form input elements with business logic.
+
+{[
+let create_ingredient name is_vegan price = ...
+
+let[@warning "-45"] pizza_schema
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/index.mld:549]: Good point, fixed.

File: sihl/src/index.mld [hunk 6]
```diff
@@ -505,6 +505,128 @@ Logs.info (fun m -> m "This allows the program user to understand what is going
...
+         (module Pizza : Rest.SERVICE with type t = Pizza.t)
+         (module View.Pizza : Rest.VIEW with type t = Pizza.t)))
+;;
+]}
+
+Don't forget to apply the CSRF and flash middlewares for the REST routes, this is a requirement. The [pizzas] router can be passed to the {!Sihl.Web.Http} service before starting the Sihl app.
+
+You have to assign the types [t] of your service and view to your resource model type.
+
+To specify which actions to support per resource, use the [only] argument. Routes of actions that are not listed return [404]
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/index.mld:611]: Fixed.

File: sihl/src/sihl.mli [hunk 7]
```diff
@@ -164,6 +164,261 @@ module Web : sig
     val set : (string * string) list -> Rock.Response.t -> Rock.Response.t
   end
 
+  (** The Rest module allow you to build RESTful web pages quickly. *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:167]: Fixed.

File: sihl/src/sihl.mli [hunk 8]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+        [name] is the schema field name and the HTML input name.
+
+        [input] is the input of the form that was submitted, it might not be
+        present. Use it to restore form values after it was submitted and
+        validation or decoding failed.
+
+        [error] is the error message of the validation or decoding. *)
+    type form = (string * string option * string option) list
+
+    (** The [SERVICE] interface has to be implemented by a CRUD service that
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:191]: Module types

File: sihl/src/sihl.mli [hunk 9]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+          You can access the original [request] directly if needed.
+
+          [csrf] token has to be included as hidden input element in the form.
+
+          [form] is the decoded and validated form from a previous request. It
+          contains input names, submitted values and error messages. This is
+          useful to display error messages on input elements or to populate the
+          form with invalid input from the failed [create] request, so the user
+          can fix it. *)
+      val new'
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:250]: The names of those actions seem to be quite popular. I would like to have `new` in the name so it doesn't break expectations of someone using rails. I can't think of a better name at them moment.

File: sihl/src/sihl.mli [hunk 10]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+        -> [> Html_types.html ] Tyxml.Html.elt Lwt.t
+
+      (** [show request resource] returns the resource instance as HTML. This is
+          the detail view of an instance of the resource. *)
+      val show
+        :  Rock.Request.t
+        -> t
+        -> [> Html_types.html ] Tyxml.Html.elt Lwt.t
+
+      (** [edit' request csrf form] returns a form to edit an instance of the
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:263]: Thanks, fixed!

File: sihl/src/sihl.mli [hunk 11]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+
+      (** [new' ?key name request] returns a form to create instances of the
+          resource as a response.
+
+          [name] is the name of the resource in plural, for example [orders] or
+          [users].
+
+          The form data is stored in the flash storage under the [key]. By
+          default, the value is [_form]. *)
+      val new'
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:305]: The names of those actions seem to be quite popular. I would like to have new in the name so it doesn't break expectations of someone using rails. I can't think of a better name at them moment.

File: sihl/src/sihl.mli [hunk 12]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+
+        A resource [pizzas] creates following 7 routers:
+
+        {v
+        GET       /pizzas          `Index  display a list of all pizzas
+        GET       /pizzas/new      `New    return an HTML form for creating a new pizza
+        POST      /pizzas          `Create create a new pizza
+        GET       /pizzas/:id      `Show   display a specific pizza
+        GET       /pizzas/:id/edit `Edit   return an HTML form for editing a pizza
+        PATCH/PUT /pizzas/:id      `Update a specific pizza
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:372]: Fixed.

File: sihl/src/sihl.mli [hunk 13]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+        DELETE    /pizzas/:id      `Delete a specific pizza
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:373]: Fixed.

File: sihl/src/sihl.mli [hunk 14]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+    module type CONTROLLER = sig
+      (** [t] is the type of the resource. *)
+      type t
+
+      (** [index name request] returns a list of all resource instances as a
+          response.
+
+          [name] is the name of the resource in plural, for example [orders] or
+          [users]. *)
+      val index : string -> Rock.Request.t -> Rock.Response.t Lwt.t
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:295]: I elaborated in the docs where it is used. I think it has to be stringly-typed yes.

File: sihl/src/sihl.mli [hunk 15]
```diff
@@ -164,6 +164,261 @@ module Web : sig
...
+      -> ('meta, 'ctor, 'resource) Conformist.t
+      -> (module CONTROLLER with type t = 'resource)
+      -> router list
+
+    (** [find_form name form] returns the [(value, error)] of a [form] input
+        element with the [name].
+
+        The [value] is the submitted value of the input element. The value is
+        set even if the submitted form failed to decode or validate. Use the old
+        valid to populate the form with old values that can be fixed by the
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/sihl.mli:414]: Fixed.

File: sihl/src/sihl.mli (hunk 1 above)
Comment (jerben) [reply sihl/src/sihl.mli:358]: The names of those actions seem to be quite popular. I would like to have new in the name so it doesn't break expectations of someone using rails. I can't think of a better name at them moment, do you have one?

File: sihl/src/web_rest.ml (hunk 2 above)
Comment (jerben) [reply sihl/src/web_rest.ml:314]: it can, fixed :)

**Comment by jerben** [comment]
@aronerben Thanks for the thorough review, I fixed the PR.

File: sihl/src/sihl.mli (hunk 9 above)
[the remark this answers is not in the document]
Comment (aronerben) [reply sihl/src/sihl.mli:250]: Ok, makes sense :)

File: sihl/src/sihl.mli (hunk 14 above)
[the remark this answers is not in the document]
Comment (aronerben) [reply sihl/src/sihl.mli:295]: Ok!

File: sihl/src/web_rest.ml [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+
+module MakeController (Service : SERVICE) (View : VIEW with type t = Service.t) =
+struct
+  exception Exception of string
+
+  type t = Service.t
+
+  let index name req =
+    let open Lwt.Syntax in
+    let csrf =
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/web_rest.ml]: Fixed.

File: sihl/src/sihl.mli (hunk 1 above)
Comment (aronerben) [reply sihl/src/sihl.mli:358]: This seems to be the best name. Only other I can think of is `remove` which doesn't convey the meaning as well IMO. Let's leave it like this then :)

File: sihl/src/web_rest.ml [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+
+let routers_of_actions
+    (type a)
+    name
+    schema
+    (module Controller : CONTROLLER with type t = a)
+    (actions : action list)
+  =
+  List.map
+    (fun action -> router_of_action (module Controller) name schema action)
```
Comment (aronerben) [inline sihl/src/web_rest.ml:303]: This can be reduced to `(router_of_action (module Controller) name schema)`, right?

Comment (jerben) [reply sihl/src/web_rest.ml]: You are right, fixed!

Comment (aronerben) [reply sihl/src/web_rest.ml]: Thanks

**Review by aronerben — APPROVED** [review]

**Pull request merged**